src/nimble.nim

```diff
@@ -72,47 +72,65 @@ proc addReverseDeps(solvedPkgs: seq[SolvedPackage], allPkgsInfo: seq[PackageInfo
         reverseDep.get.isLink = true
       addRevDep(options.nimbleData, solvedPkg.get.basicInfo, reverseDep.get)
 
-proc processFreeDependenciesSAT(rootPkgInfo: PackageInfo, pkgList: seq[PackageInfo], options: Options): HashSet[PackageInfo] = 
+var satProccesedPackages: HashSet[PackageInfo]
+proc processFreeDependenciesSAT(rootPkgInfo: PackageInfo, options: Options): HashSet[PackageInfo] = 
+  if satProccesedPackages.len > 0:
+    return satProccesedPackages
   var solvedPkgs = newSeq[SolvedPackage]()
   var pkgsToInstall: seq[(string, Version)] = @[]
+  var pkgList = initPkgList(rootPkgInfo, options)
   var allPkgsInfo: seq[PackageInfo] = pkgList & rootPkgInfo
+  var rootPkgInfo = rootPkgInfo
+  #Replace requirements so they are updated as needed 
+  if options.action.typ == actionUpgrade:
+    let toUpgradeNames = options.action.packages.mapIt(it[0])
+    pkgList = pkgList.filterIt(it.basicInfo.name notin toUpgradeNames)
+
+    # rootPkgInfo.requires = rootPkgInfo.requires.filterIt(it.name notin toUpgradeNames)
+    # rootPkgInfo.requires &= options.action.packages
 
   result = solveLocalPackages(rootPkgInfo, pkgList, solvedPkgs)
   if solvedPkgs.len > 0: 
     displaySatisfiedMsg(solvedPkgs, pkgsToInstall)
     addReverseDeps(solvedPkgs, allPkgsInfo, options)
-    return result
+    for pkg in allPkgsInfo:
+      result.incl pkg
+    result = 
+      result.toSeq
+      .deleteStaleDependencies(rootPkgInfo, options)
+      .deduplicate.toHashSet
```

Comment:
Quite sure that `toHashSet` does the deduplicate step, and more efficiently too.